src/ucp/core/ucp_worker.c

```diff
@@ -3066,6 +3072,13 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
 
         ucp_worker_keepalive_next_ep(worker);
         ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
+
+        if (worker->keepalive.iter == worker->keepalive.iter_begin) {
```

Comment:
shouldn't it be outside the if (from line 3069)?

Reply:
the idea is to check the new iterator instead of the previous one

Reply:
but can't this situation happen also if the iterator is not exactly list head? 

> the idea is to check the new iterator instead of the previous one

we can check after this if..

Reply:
discussed offline, just need to update the commend and force push